src/registry/AvatarRegistry.sol

```diff
@@ -0,0 +1,523 @@
+            revert NotKeeper(msg.sender);
+        }
+        _;
+    }
+
+    /// @dev Fallback function accepts Ether transactions.
+    receive() external payable {
+        emit RegistryEthReceived(msg.sender, msg.value);
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // PUBLIC: Governance
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @dev It will initiate the upKeep job for monitoring avatars
+    /// @notice only callable via governance
+    /// @param gasLimit gas limit for the avatar monitoring upkeep task
+    function initializeBaseUpkeep(uint256 gasLimit) external onlyGovernance {
+        if (gasLimit == 0) {
+            revert ZeroUintValue();
+        }
+
+        avatarMonitoringUpKeepId = _registerUpKeep(address(this), gasLimit, AVATAR_REGISTRY_NAME);
+
+        if (avatarMonitoringUpKeepId > 0) {
+            /// @dev give allowance of spending LINK funds
+            LINK.approve(address(CL_REGISTRY), type(uint256).max);
+        }
+    }
+
+    /// @dev Adds an avatar into the registry
+    /// @notice only callable via governance
+    /// @param avatarAddress contract address to be register as new avatar
+    /// @param name avatar's name
+    /// @param gasLimit amount of gas to provide the target contract when
+    /// performing upkeep
+    function addAvatar(address avatarAddress, string memory name, uint256 gasLimit) external onlyGovernance {
+        /// @dev sanity checks before adding a new avatar in storage
+        if (avatarAddress == address(0)) {
+            revert ZeroAddress();
+        }
+        if (avatarsInfo[avatarAddress].gasLimit != 0) {
+            revert AvatarAlreadyRegister(avatarAddress);
+        }
+        if (gasLimit == 0) {
+            revert ZeroUintValue();
+        }
+        if (bytes(name).length == 0) {
+            revert EmptyString();
+        }
+        if (IAvatar(avatarAddress).keeper() != address(CL_REGISTRY)) {
+            revert NotCLKeeperSet();
+        }
+
+        _avatars.add(avatarAddress);
+        avatarsInfo[avatarAddress] = AvatarInfo({
+            name: name,
+            gasLimit: gasLimit,
+            status: AvatarStatus.TESTING,
+            upKeepId: _registerUpKeep(avatarAddress, gasLimit, name)
+        });
+
+        emit NewAvatar(avatarAddress, name, gasLimit, block.timestamp);
+    }
+
+    /// @dev Cancels an avatar upkeep job
+    /// @notice only callable via governance
+    /// @param avatarAddress contract address to be cancel upkeep
+    function cancelAvatarUpKeep(address avatarAddress) external onlyGovernance {
+        if (!_avatars.contains(avatarAddress)) {
+            revert NotAvatarIncluded(avatarAddress);
+        }
+        // NOTE: only avatar which upkeep is being cancelled can be removed
+        uint256 upKeepId = avatarsInfo[avatarAddress].upKeepId;
+        CL_REGISTRY.cancelUpkeep(upKeepId);
+    }
```

Comment:
would this not also require some adjustments in the registry? or is this done manually by calling `updateStatus`? why the separation?